### ISSUE TITLE
Cannot Grab cards out of opponents library

### ISSUE THREAD

**Issue by tehboss** [Cockatrice/Cockatrice#625]
A player can no longer grab cards out of opponents library.  This used to work.

**Comment by Daenyth** [comment]
@poixen I bet this is related to pile view. This breaks several cards

**Comment by Daenyth** [comment]
@tehboss Do you happen to know what version worked for this? I remember it working, just trying to track down where it broke

**Comment by Daenyth** [comment]
@tehboss What operating system are you on?

**Comment by tehboss** [comment]
Windows 7 x64, and no, It’s been a bit.  I know it worked sometime before the play screen went to purple.  Short of be going back through my downloads and attempting the installs, I couldn’t guess.

**Comment by ctrlaltca** [comment]
I suppose this is caused by #428.

**Comment by Daenyth** [comment]
That seems likely based on the description

**Comment by ZeldaZach** [comment]
Maybe we can add a new action to the server `"Foo has closed your library"` so people will know when it's closed, and then we can revert #428

**Comment by Daenyth** [comment]
Summarizing from some gitter chat; It was broken because the old view allowed you to see what cards your opp was drawing as it live updated.

I think the fix is to permit non-owners to drag out of a revealed deck, but only if the deck is in the same state that it was in before the reveal - if it's been shuffled or modified in any way the steal needs to be rejected and the client needs to be instructed to ask the owner for a new view

**Comment by ZeldaZach** [comment]
Have we thought of a way to validate a deck configuration? Is it possible we can gen a quick hash based on the order of the cards and compare them at that point?

**Comment by Daenyth** [comment]
It wouldn't be hard

**Comment by ashleytrash** [comment]
why hasn't this been fixed yet?

**Comment by ZeldaZach** [comment]
@ashleytrash We haven't come up with a solution that doesn't ruin game play yet, we still need to have a plan before we can do anything.

**Comment by tehboss** [comment]
How exactly does allowing an opponent the ability to grab a card from an opponent’s library ruin gameplay? It seems worse to me when playing certain cards that if an opponent needs to choose a card and you shouldn’t see it (Sphinx Ambassador is a prime example of play that can’t happen due to this restriction)

**Comment by ZeldaZach** [comment]
When you give them a live version of your deck, it edits in real time. If they choose not to close the library and you continue the game, they can see actions you take. For example, if you draw a card, another card will be removed from the dialog (the one you drew) so they can infer what you drew. We found this to be a bigger issue then the selecting from library.

If you have any proposed solutions, please do let us know

**Comment by tehboss** [comment]
Sure. I was thinking that you can have a sub menu that you are allowing the other player to select a card or more. With that routine open, a notification is sent to all players that player A is looking at player B’s library, and when Player A closes the library and auto-shuffles for Player B. Then everyone knows and all can wait until the choice is finished.

**Comment by cataclysam** [comment]
Hi - I really would love a fix for this. It seems like a simple one too. It doesn't have to be perfect, just send a message in chat when I stop looking, as with looking at your own library. Then the opponent can just not do anything until it's closed.

**Comment by ebbit1q** [comment]
I made a writeup of solid workaround for cockatrice that will make sure no information gets leaked, making use of newer features and keybinds:
1. hold ctrl+shift+e to move your deck to play face down (I'd like an alternative or this)
2. your opponent moves their hand to play facedown by dragging to select all of them and put them on their board, you verify that their hand is empty
3. drag over the cards you just moved to play to select all of them, then drag them to your opponent's board
4. your opponent picks up all the cards you placed on their board to their hand and can use the "view hand" option to search through it
5. your opponent chooses a card and places it on their field face down by dragging it and holding shift
6. your opponent selects all their cards in their hand by dragging, then moves them to play face down by holding shift, you verify that their hand is empty before they pick up their hand again
7. your opponent moves your deck from their field to yours and you drag it back to your library afterwards
8. you have to press ctrl+s to shuffle, because the randomness of your library could be influenced by your opponent otherwise!

this should allow you to resolve praetor's grasp, portent, gonti, what have you while maintaining rules compliance

in order to not confuse the three different piles of facedown cards (your library, opponent's hand, picked cards) I propose you use different lines on your field for them

these steps are fairly fragile though, either player (mostly the active player) can fuck them up by forgetting to keep the cards face down

**Comment by Edahsrevlis** [comment]
Proposal #1: If the ability to 'Play Card Face Down' from an opponent's library were enabled, you could make that command also close the dialog, so that only 1 card can be played before the library is no longer revealed. Any effect that requires multiple searches would just need the library revealed multiple times. However, I play a deck that runs Sadistic Sacrament and Neverending Torment, so this option would still be fairly tedious.

Proposal #2: Run a check before all deck-related actions like scry, draw, shuffle, move top cards to graveyard, move top cards to exile... before any of those actions, call a function that checks all active players and forces them to cease viewing your library and reshuffle it at that moment, then do the action (draw). Is it possible for my shuffle to force everyone else to stop looking at my library? Or does Cockatrice use OS dialog boxes and therefore not have enough control over them?

Proposal #3: If the deck is being searched and hasn't been closed, keep a boolean that says Deck.isBeingViewedByOthers, and block all deck-related actions and give a warning message in chat until it has been closed.

Proposal #4: Option to Create Pile... which will be a visual object the size of a normal card. A pile with 0 cards is a black rectangle with a white number counter like the deck, graveyard, and sideboard have. A pile is face-down like a library.  It has these features:
- move Pile to position
- drag card from zone to Pile
- give Pile to player
- shuffle Pile
- destroy Pile (any existing Pile cards are played face-down)
- draw top card from pile
- view Pile (can interact as though it's a library, moving cards to other zones)

Option #4 may require the most work, but it would also solve currently unsupported cards like Parallel Thoughts, which require shuffling a miscellaneous pile. This could prevent a lot of dancing around between existing zones and visual board spam by creating any number of small, neat Pile objects.

**Comment by ebbit1q** [comment]
Your proposals are great but you need to see this from a server side view, the current show deck command gives a user a full view of the deck that stays live, it needs an additional option to either deny this or give a different type of command. Client side fixes to this are not going to work, the server has no control over these.

**Comment by ebbit1q** [comment]
in competitive settings where judges are available we now have an alternative since the inclusion of https://github.com/Cockatrice/Cockatrice/pull/4297

in this example; you = the player resolving praetor's grasp; opponent = the player targeted by praetor's grasp; judge = a third party with judge permissions which will facilitate the entire action with respect to preservation of game information
1. you call a judge
2. judge joins your game, judges join games as spectator and in judge mode by holding shift while joining
3. judge selects "reveal library to your player name" on your opponent
3. you choose a card and send a private message to the judge with the desired card (preferably you should copy it by clicking it while holding alt then cutting this message with ctrl+x and sending it to the judge but this is the only part liable to player error)
4. judge selects "view library" on the opponent, this will no longer show the library to opponent
5. judge drags the card face down onto opponent's field by holding shift
6. judge closes the view library window and shuffles opponent's library
6. judge drags the card to your field
7. you verify the correct card has been given by right clicking the card and selecting "peek at card face"
8. judge leaves

if at step 8 the wrong card has been provided by the judge, the judge should:
1. move the wrong card back to opponent's field
2. move the wrong card to opponent's deck
3. shuffle opponent's deck
4. continue from step 5 above (select view library)

note that you can keep open the read only view of your opponent's library (this has no effect on gameplay and means there is no incentive for players to take a screenshot of it)

this method means judges will perform all of the actions, preventing players from making mistakes in tournament settings where we might have something on the line.
in total you will perform 3 actions (call judge, pm judge, peek at card), opponent does nothing and the judge does 7, this means judges are encouraged to practice this action once in case of a tournament.

**Comment by n21lv** [comment]
> I made a writeup of solid workaround for cockatrice that will make sure no information gets leaked, making use of newer features and keybinds:
> 
> 1. hold ctrl+shift+e to move your deck to play face down (I'd like an alternative or this)
> 2. your opponent moves their hand to play facedown by dragging to select all of them and put them on their board, you verify that their hand is empty
> 3. drag over the cards you just moved to play to select all of them, then drag them to your opponent's board
> 4. your opponent picks up all the cards you placed on their board to their hand and can use the "view hand" option to search through it
> 5. your opponent chooses a card and places it on their field face down by dragging it and holding shift
> 6. your opponent selects all their cards in their hand by dragging, then moves them to play face down by holding shift, you verify that their hand is empty before they pick up their hand again
> 7. your opponent moves your deck from their field to yours and you drag it back to your library afterwards
> 8. you have to press ctrl+s to shuffle, because the randomness of your library could be influenced by your opponent otherwise!
> 
> this should allow you to resolve praetor's grasp, portent, gonti, what have you while maintaining rules compliance
> 
> in order to not confuse the three different piles of facedown cards (your library, opponent's hand, picked cards) I propose you use different lines on your field for them
> 
> these steps are fairly fragile though, either player (mostly the active player) can fuck them up by forgetting to keep the cards face down

I had come up with what I think is an upgraded version of this yesterday, that also solves almost all trust issues.
1. Reveal top N (where N equals number of cards left in your library) in your library to the player who needs to grab a card from your library.
2. Cockatrice will display a small window to the chosen player, with all the cards in your library nicely ordered and split into colulms as evenly as possible. With this information and some basic maths, a player can figure out the relative position of any card in the deck. E.g. if you have 80 cards left in your deck, they will be split into into five 14-card-columns and one with 10 cards in it. So, your opponent will have to find which column has the card and remember its relative position (let's call this number P), assuming that counting starts from the top card in the leftmost column. 
3. Then you press Ctrl+Shift+E once to figure out what number will Cockatrice assign to the first face-down card you play (they don't reset each time so you can't figure out if the same card gets played face-down multiple times). Let's call that number X.
4. Then you hold Ctrl+Shift+E until a face-down card with number X+P-1 hits the board.
5. Pass that card to the opponent.
6. Opponent checks if that's the right card and puts it to their hand.
7. If that card had to be exiled face-down, opponent plays the card face down, so it becomes a new object that its original owner can't peek at (Cockatrice seems to allow that for some reason even after you move it to the opponent's board, but I only tested this solution in a local game).
8. You put all the face-down cards you played from your library in steps 3 and 4 on top of your library and shuffle it.

**Comment by ebbit1q** [comment]
not sure if this is less complicated but sounds like it'd work just as well, I think you'd still need to shuffle before showing your deck though, I think that is required to make the card ids line up

**Comment by ebbit1q** [comment]
also, show top n is the same as show deck and disabling sorting of any kind

**Comment by n21lv** [comment]
> also, show top n is the same as show deck and disabling sorting of any kind

I know, but it requires some extra clicks (which players can forget about). 
Try testing this solution in a local game, it seems to work really well for me.
